middleware.go

```diff
@@ -0,0 +1,110 @@
+package jwt
+
+import (
+	"github.com/Digital-Insight-Technologies-Ltd/gin-jwt-cognito/models"
+	"github.com/sirupsen/logrus"
+	"net/http"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/go-errors/errors"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/google/uuid"
+)
+
+// UserClaimsKey is the key for user claims in our context
+const UserClaimsKey = "userClaims"
+
+// CognitoClaimsMiddleware is a middleware for Gin that extracts user claims from
+// AWS Cognito.
+func CognitoClaimsMiddleware() gin.HandlerFunc {
+	// parser
+	jwtParser := new(jwt.Parser)
+
+	return func(c *gin.Context) {
+		authHeader := c.GetHeader("Authorization")
+		if !strings.HasPrefix(authHeader, "Bearer ") {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Missing or invalid Authorization header"})
+			return
+		}
+		claims, err := parseClaims(jwtParser, authHeader[7:])
+
+		if err != nil {
+			logrus.WithError(err).Error("Error parsing claims (OIDC: stage 1)")
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 1)"})
+			return
+		}
+
+		userCtx, err := constructUserContext(claims)
+
+		if err != nil {
+			logrus.WithError(err).Error("Error constructing user context (OIDC: stage 2)")
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 2)"})
+			return
+		}
+
+		// Add the claims to Gin's context
+		c.Set(UserClaimsKey, userCtx)
+
+		// Proceed with the middleware chain
+		c.Next()
+	}
+}
+
+func parseClaims(jwtParser *jwt.Parser, claimsHeader string) (jwt.MapClaims, error) {
+	tok, _, err := jwtParser.ParseUnverified(claimsHeader, jwt.MapClaims{})
+	if err != nil {
+		logrus.WithError(err).Error("Error parsing claims")
+		return nil, err
+	}
+
+	claims, ok := tok.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, errors.New("invalid claims type")
```

Comment:
Consider using standard error handling mechanisms provided by Go instead of third-party error packages like `github.com/go-errors/errors` to reduce dependencies and improve performance. [medium]